core/src/Cuda/Kokkos_Cuda_Vectorization.hpp

```diff
@@ -48,347 +48,165 @@
-}
-
-template <typename Scalar>
-KOKKOS_INLINE_FUNCTION Scalar shfl_down(
-    const Scalar& val, const int& delta,
-    const typename Impl::enable_if<(sizeof(Scalar) == 8), int>::type& width) {
-  int lo           = __double2loint(*reinterpret_cast<const double*>(&val));
-  int hi           = __double2hiint(*reinterpret_cast<const double*>(&val));
-  lo               = KOKKOS_IMPL_CUDA_SHFL_DOWN(lo, delta, width);
-  hi               = KOKKOS_IMPL_CUDA_SHFL_DOWN(hi, delta, width);
-  const double tmp = __hiloint2double(hi, lo);
-  return *(reinterpret_cast<const Scalar*>(&tmp));
-}
+// Include all lanes
+constexpr unsigned shfl_all_mask = 0xffffffff;
 
-#endif  // ( CUDA_VERSION < 9000 )
+}  // end namespace Impl
 
-template <typename Scalar>
-KOKKOS_INLINE_FUNCTION Scalar shfl_down(
-    const Scalar& val, const int& delta,
-    const typename Impl::enable_if<(sizeof(Scalar) == 4), int>::type& width) {
-  Scalar tmp1 = val;
-  float tmp   = *reinterpret_cast<float*>(&tmp1);
-  tmp         = KOKKOS_IMPL_CUDA_SHFL_DOWN(tmp, delta, width);
-  return *reinterpret_cast<Scalar*>(&tmp);
-}
-
-template <typename Scalar>
-KOKKOS_INLINE_FUNCTION Scalar shfl_down(
-    const Scalar& val, const int& delta,
-    const typename Impl::enable_if<(sizeof(Scalar) > 8), int>::type& width) {
-  Impl::shfl_union<Scalar> s_val;
-  Impl::shfl_union<Scalar> r_val;
-  s_val = val;
-
-  for (int i = 0; i < s_val.n; i++)
-    r_val.fval[i] = KOKKOS_IMPL_CUDA_SHFL_DOWN(s_val.fval[i], delta, width);
-  return r_val.value();
-}
+namespace Impl {
 
-KOKKOS_INLINE_FUNCTION
-int shfl_up(const int& val, const int& delta, const int& width) {
-  return KOKKOS_IMPL_CUDA_SHFL_UP(val, delta, width);
-}
+//----------------------------------------------------------------------------
+// Shuffle operations require input to be a register (stack) variable
+
+// Derived implements do_shfl_op(unsigned mask, T& in, int lane, int width),
+// which turns in to one of KOKKOS_IMPL_CUDA_SHFL(_UP_|_DOWN_|_)MASK
+// Since the logic with respect to value sizes, etc., is the same everywhere,
+// put it all in one place.
+template <class Derived>
+struct in_place_shfl_op {
+  // CRTP boilerplate
+  __device__ __inline__ const Derived& self() const noexcept {
+    return *static_cast<Derived const*>(this);
+  }
 
-KOKKOS_INLINE_FUNCTION
-float shfl_up(const float& val, const int& delta, const int& width) {
-  return KOKKOS_IMPL_CUDA_SHFL_UP(val, delta, width);
-}
+  // sizeof(Scalar) == sizeof(int) case
+  template <class Scalar>
+  // requires _assignable_from_bits<Scalar>
+  __device__ inline typename std::enable_if<sizeof(Scalar) == sizeof(int)>::type
+  operator()(Scalar& out, Scalar const& in, int lane_or_delta, int width,
+             unsigned mask = shfl_all_mask) const noexcept {
+    //------------------------------------------------
+    reinterpret_cast<int&>(out) = self().do_shfl_op(
+        mask, *reinterpret_cast<int const*>(&in), lane_or_delta, width);
```

Comment:
```suggestion
        mask, reinterpret_cast<int const&>(in), lane_or_delta, width);
```
(You may ignore this)